packages/https-proxy/lib/server.coffee

```diff
@@ -69,33 +70,66 @@ class Server
       res.end()
     .pipe(res)
 
+  _needsUpstreamProxy: (hostname, port) ->
+    process.env.HTTP_PROXY
+
   _makeDirectConnection: (req, socket, head) ->
     { port, hostname } = url.parse("http://#{req.url}")
 
+    debug(process.env.HTTP_PROXY)
+
+    if @_needsUpstreamProxy(hostname, port)
+      upstreamProxy = process.env.HTTP_PROXY
+      return @_makeUpstreamProxyConnection(upstreamProxy, socket, head, port, hostname)
+
+    debug("Making direct connection to #{hostname}:#{port}")
     @_makeConnection(socket, head, port, hostname)
 
   _makeConnection: (socket, head, port, hostname) ->
-    cb = ->
+    onConnect = ->
       socket.pipe(conn)
       conn.pipe(socket)
       socket.emit("data", head)
 
       socket.resume()
 
     ## compact out hostname when undefined
-    args = _.compact([port, hostname, cb])
+    args = _.compact([port, hostname, onConnect])
 
     conn = net.connect.apply(net, args)
 
     conn.on "error", (err) =>
       if @_onError
         @_onError(err, socket, head, port)
 
+  _makeUpstreamProxyConnection: (upstreamProxy, socket, head, toPort, toHostname) ->
+    debug("making proxied connection to #{toHostname}:#{toPort} with upstream #{upstreamProxy}")
+
+    if not @httpsAgents[upstreamProxy]
+      @httpsAgents[upstreamProxy] = new httpsAgent(upstreamProxy)
+
+    agent = @httpsAgents[upstreamProxy]
```

Comment:
need to do more testing to understand whether this is the right thing to do, and how it impacts performance